src/redisearch_rs/buffer/src/reader.rs

```diff
@@ -0,0 +1,93 @@
+use crate::Buffer;
+
+/// A cursor to read from a [`Buffer`].
+///
+/// It's best to use this through the `std::io::Read` implementation.
+///
+/// # Invariants
+///
+/// 1. Position is smaller than or equal to the length of the buffer we're reading from.
+/// 2. `BufferReader` has the same memory layout of [`ffi::BufferReader`].
+#[repr(C)]
+pub struct BufferReader<'a> {
+    buffer: &'a Buffer,
+    position: usize,
+}
+
+impl<'a> BufferReader<'a> {
+    /// Create a new cursor, reading from the beginning of the buffer.
+    pub fn new(buffer: &'a Buffer) -> Self {
+        Self {
+            buffer,
+            position: 0,
+        }
+    }
+
+    /// Create a new cursor, reading from the given position inside the buffer.
+    ///
+    /// # Panic
+    ///
+    /// Panics if `position` is out of bounds—i.e. if it's greater than the
+    /// buffer offset.
+    pub fn new_at(buffer: &'a Buffer, position: usize) -> Self {
+        if position > buffer.0.offset {
+            panic!(
+                "Position out of bounds: {} is greater than the buffer offset {}",
+                position, buffer.0.offset
+            );
+        }
+        Self { buffer, position }
+    }
+
+    /// The current position of the reader.
+    pub fn position(&self) -> usize {
+        self.position
+    }
+
+    /// A reference to the buffer we're reading from.
+    pub fn buffer(&self) -> &Buffer {
+        &self.buffer
+    }
+}
+
+impl<'a> std::io::Read for BufferReader<'a> {
+    fn read(&mut self, mut dest_buf: &mut [u8]) -> std::io::Result<usize> {
+        use std::io::Write as _;
+
+        debug_assert!(self.position <= self.buffer.0.offset);
+        // No out-of-bounds risk, thanks to `BufferReader`'s invariant 1.
+        let n_bytes = dest_buf.write(&self.buffer.as_slice()[self.position..])?;
+        self.position += n_bytes;
+
+        Ok(n_bytes)
+    }
+}
+
+// Check, at compile-time, that `BufferReader` and `ffi::BufferReader` have the same representation.
+const _: () = {
```

Comment:
why not an ignored test instead?

Reply:
That's not what I meant. :) I didn't know that either.

I was talking about:

```rust
#[test]
#[ignore]
```

which tests the code to compile but not actually run.

Reply:
While that would be technically correct (i.e. the current test desugars to a no-op at runtime), I think it'd be confusing to see that `#[ignore]` for folks who are not too well-versed into the semantics of const evaluation. 
Since the penalty is negligible, I prefer to leave it as is.